### ISSUE TITLE
Support icons in build order steps

### ISSUE THREAD

**Issue by jensbuehl** [jensbuehl/aoe4-guides#16]
E.g. use villager icon instead of plain text in the step description field.
Inspired by: https://age4builder.com/

**Comment by CraftySalamander** [comment]
Suggestion to make the task easier for Overlay export.

You can copy [this folder](https://github.com/CraftySalamander/RTS_Overlay/tree/master/pictures/aoe4) with all the images, without changing its architecture (same images for AoE4 Overlay & RTS Overlay, see my explanation in the overlay export issue).

In the overlay, each image is indicated as its path relative to this folder. For instance, "@unit_worker/villager.png@" will look for the image[ "villager.png"](https://github.com/CraftySalamander/RTS_Overlay/blob/master/pictures/aoe4/unit_worker/villager.png) in the [folder "unit_worker"](https://github.com/CraftySalamander/RTS_Overlay/tree/master/pictures/aoe4/unit_worker).

If you use these images without any change in the architecture, adapting the overlay export (and potentially the overlay JSON import function) will be super easy because the names and paths will be identical.

In case you want to use your own images and paths, it can still work, but you will need to define a dictionary (like [this one](https://github.com/LENpolygon/Build-Order-Tool-AoE4-/blob/main/public/js/AoE4_Overlay.js#L12)).

**Comment by jensbuehl** [comment]
Great, thanks for your valuable input!

**Comment by jensbuehl** [comment]
Current state of imported images from AoE4_Overlay:
![Screenshot 2023-04-13 135357](https://user-images.githubusercontent.com/3983913/231750801-87a2b62b-b820-47af-9293-19817427a940.png)

TODO: Editor + optional tooltips

**Comment by jensbuehl** [comment]
@CraftySalamander While adapting the code for icon support, I found at least one missing image. How did you grab these images? (without background to be consistent)

So far I noticed that the Fertile Crescent Abbasid technology is missing.

**Comment by CraftySalamander** [comment]
I found them on aoe4world, see for instance https://aoe4world.com/explorer/civs/abbasid/technologies/fertile-crescent

This one makes sense since it was not part of the game when I made the last image update.

Once you are done adding all the missing images, can you give me the list (with folders) of new images?
I will then be able to add them on RTS Overlay (and AoE4 Overlay if fine for Maguro)?

**Comment by jensbuehl** [comment]
Sure, will keep you posted so that the images stay in sync!

**Comment by CraftySalamander** [comment]
FYI, I prepared the code update in RTS Overlay to link to aoe4guides.com (see [here](https://github.com/CraftySalamander/RTS_Overlay/commit/1c06a6f3e83873002e93f6eb3f871f9ac0aa628c#diff-181bb32a3d7178143937c386d4e49e84cda5543c3bf6252a3e089f3da220a81b)).
It will be part of release 1.5.1.

I will suggest the same to Maguro for AoE4 Overlay once the illustration format is ready.

**Comment by jensbuehl** [comment]
Illustrations will go live today.
Just pull the pictures folder and diff it on your end - should include all (or most) of the Season 4 changes.
https://github.com/jensbuehl/aoe4-guides/tree/main/public/assets/pictures

**Comment by CraftySalamander** [comment]
Very nice!

RTS Overlay update, see [here](https://github.com/CraftySalamander/RTS_Overlay/commit/9ad84e2f10c805f4b79ec7cd6ac41caea0fee45b).

AoE4 Overlay pull request, see [here](https://github.com/FluffyMaguro/AoE4_Overlay/pull/15).

**Comment by CraftySalamander** [comment]
@jensbuehl 
There seems to be a bug when using the `Overlay Tool` copy function.

[This BO](https://aoe4guides.com/builds/52mdvDz14wPxsvMcMnzz) export seems perfectly fine, I get notes like this:
`"5 @unit_worker/villager.png@ to @resource/sheep.png@, 1 @unit_worker/villager.png@ to build a @building_economy/house.png@ and a @building_economy/mill.png@ (queue vills)"`

But [this BO](https://aoe4guides.com/builds/kR15ALIOofeNFRoO8Lqb) export seems buggy, I get notes like this:
`"Split the first 6&nbsp;<img src=\"https://aoe4guides.com/assets/pictures/unit_worker/villager.png\" class=\"icon\">between&nbsp;<img src=\"https://aoe4guides.com/assets/pictures/resource/resource_food.png\" class=\"icon\">(1)&nbsp;<img src=\"https://aoe4guides.com/assets/pictures/resource/resource_stone.png\" class=\"icon\">(5)."`

Additional point, it is `"builder"` and not `"builders"` (no `s`) to use in `"resources"`.

**Comment by jensbuehl** [comment]
@CraftySalamander Thanks for the feedback and testing. 
builders > builder fixed locally. Will deploy soonish.

The export issue is related to preserved empty spaces that slip in when e.g. multiple spaces in a row or using line breaks. So the spaces are really part of the data and exported "correctly". 

I have to take a look - might slip into next week, though.
Idea: Remove these redundant spaces when exporting. (string.replace(...))
Idea: Filter

**Comment by jensbuehl** [comment]
Can be verified using this BO on dev:
https://aoe4-guides-dev.web.app/builds/gADkt4R7YvqSAisUDPVx

(Line 1 is exported fine. Line 2 is broken due to the preserved spaces.)

**Comment by jensbuehl** [comment]
There was a second issue regarding the <img> format. Did not work when closing tag was missing.
I fixed both. Both BOs work fine in the Overlay tool now.

Greetings and have a nice weekend :)

**Comment by CraftySalamander** [comment]
Great :-)
Enjoy the weekend!
